homeassistant/components/google_assistant/trait.py

```diff
@@ -750,3 +751,188 @@ def query_attributes(self):
                 ATTR_ENTITY_ID: self.state.entity_id,
                 fan.ATTR_SPEED: params['fanSpeed']
             }, blocking=True)
+
+
+@register_trait
+class ModesTrait(_Trait):
+    """Trait to set modes.
+
+    https://developers.google.com/actions/smarthome/traits/modes
+    """
+
+    name = TRAIT_MODES
+    commands = [
+        COMMAND_MODES
+    ]
+
+    # Google requires specific mode names and settings. Here is the full list.
+    # https://developers.google.com/actions/reference/smarthome/traits/modes
+    # All settings are mapped here as of 2018-11-28 and can be used for other
+    # entity types.
+
+    HA_TO_GOOGLE = {
+        media_player.ATTR_INPUT_SOURCE: "input source",
+    }
+    SUPPORTED_MODE_SETTINGS = {
+        'xsmall': [
+            'xsmall,', 'extra', 'small,', 'min,', 'minimum,', 'tiny,', 'xs'],
+        'small': ['small,', 'half'],
+        'large': ['large,', 'big,', 'full'],
+        'xlarge': ['extra', 'large,', 'xlarge,', 'xl'],
+        'Cool': ['cool,', 'rapid', 'cool,', 'rapid', 'cooling'],
+        'Heat': ['heat'], 'Low': ['low'],
+        'Medium': ['medium,', 'med,', 'mid,', 'half'],
+        'High': ['high'],
+        'Auto': ['auto,', 'automatic'],
+        'Bake': ['bake'], 'Roast': ['roast'],
+        'Convection Bake': ['convection', 'bake,', 'convect', 'bake'],
+        'Convection Roast': ['convection', 'roast,', 'convect', 'roast'],
+        'Favorite': ['favorite'],
+        'Broil': ['broil'],
+        'Warm': ['warm'],
+        'Off': ['off'],
+        'On': ['on'],
+        'Normal': [
+            'normal,', 'normal', 'mode,', 'normal', 'setting,', 'standard,',
+            'schedule,', 'original,', 'default,', 'old', 'settings'
+        ],
+        'None': ['none'],
+        'Tap Cold': ['tap', 'cold'],
+        'Cold Warm': ['cold', 'warm'],
+        'Hot': ['hot'],
+        'Extra Hot': ['extra', 'hot'],
+        'Eco': ['eco'],
+        'Wool': ['wool,', 'fleece'],
+        'Turbo': ['turbo'],
+        'Rinse': ['rinse,', 'rinsing,', 'rinse', 'wash'],
+        'Away': ['away,', 'holiday'],
+        'maximum': ['maximum'],
+        'media player': ['media', 'player'],
+        'chromecast': ['chromecast'],
+        'tv': [
+            'tv,', 'television,', 'tv', 'position,', 'television', 'position,',
+            'watching', 'tv,', 'watching', 'tv', 'position,', 'entertainment,',
```

Comment:
I guess these duplicated values aren't supposed to be here? Also, why do some entries have an comma appended?

Reply:
Fixed. Did a <table> to JSON rip and it apparently had a hard time with that.
I also apparently had a hard time reading.